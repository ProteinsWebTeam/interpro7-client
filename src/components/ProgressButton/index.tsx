import React from 'react';

import cssBinder from 'styles/cssBinder';

import local from './style.css';
import fonts from 'EBI-Icon-fonts/fonts.css';

const css = cssBinder(local, fonts);

const CENTER = 90;
const RADIUS = 70;
const STROKE_WIDTH = CENTER - RADIUS;
const CIRCUMFERENCE = 2 * Math.PI * RADIUS;

type Props = {
  downloading: boolean;
  success?: boolean;
  failed: boolean;
  progress: number;
  showIcon?: boolean;
  iconType?: string;
};

const ProgressButton = ({
  downloading,
  success,
  progress,
  failed,
  showIcon = true,
}: Props) => {
  const iconClass = success ? 'icon-save' : 'icon-download-alt';

  return (
    <>
      <span className={css('container', { failed })}>
        <svg
          width="2em"
          height="2em"
          viewBox={`0 0 ${2 * CENTER} ${2 * CENTER}`}
          className={css('svg')}
        >
          <g className={css('group', { downloading })}>
            <circle
              cx={CENTER}
              cy={CENTER}
              transform={`rotate(-90 ${CENTER} ${CENTER})`}
              r={RADIUS}
              opacity={downloading ? 1 : 0}
              fill="none"
              className={css('progress')}
              strokeWidth={STROKE_WIDTH}
              strokeLinecap="round"
              strokeDasharray={CIRCUMFERENCE}
              strokeDashoffset={CIRCUMFERENCE * (1 - progress)}
            />
            <circle
              cx={CENTER}
              cy={CENTER}
              // circle inside the progress circle (+1 to account for interpolat.)
              r={RADIUS - STROKE_WIDTH / 2 + 1}
              fill="none"
            />
          </g>
        </svg>
        {showIcon && (
          <span
            className={css('download-icon', 'icon', 'icon-common', iconClass)}
          />
        )}
      </span>
    </>
  );
};

export default ProgressButton;
