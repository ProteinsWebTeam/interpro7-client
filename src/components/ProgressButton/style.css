@import "../../styles/timing.css";
@import "../../styles/colors.css";

.svg {
  margin: auto;
}

.container {
  display: inline-block;
  line-height: 0;
}

.failed {
  cursor: not-allowed;
}

@keyframes pulsing-rotate {
  0% {
    transform: rotate(0) scale(1);
  }

  33% {
    transform: rotate(-180deg) scale(0.95);
  }

  100% {
    transform: rotate(-360deg) scale(1);
  }
}

.group {
  transform-origin: center;
  transform: scale(1);

  &.downloading {
    animation-duration: var(--timing-slow);
    animation-timing-function: linear;
    animation-delay: var(--timing-medium);
    animation-iteration-count: infinite;
    animation-direction: reverse;
    animation-name: pulsing-rotate;
  }
}

.shadow {
  transform-origin: center;
  transition: transform var(--timing-medium) ease-in-out;
  transform: scale(0.8);
  stroke: var(--colors-main);
  opacity: 0.2;

  &.downloading {
    transform: scale(1);
  }
}

.progress {
  transition: stroke-dashoffset var(--timing-instant) ease-in-out;
  stroke: var(--colors-progress); /*sat. 8ecddc*/
}

.icon-flip {
  perspective: 500px;

  & > * {
    backface-visibility: hidden;
    transform-origin: center;
    transition: transform var(--timing-fast) linear;
  }

  &.flipped {
    & > .front {
      transform: rotateY(90deg);
      transition-delay: 0s;
    }

    & > .back {
      transform: rotateY(0);
      transition-delay: var(--timing-fast);
    }
  }
}

.front {
  transform: rotateY(0);
  transition-delay: var(--timing-fast);
}

.back {
  transform: rotateY(90deg);
  transition-delay: 0s;
}

.svg-overlap {
  position: absolute;
  top: 50%;
  margin-top: -8px;
  left: 50%;
  margin-left: -8px;
}
