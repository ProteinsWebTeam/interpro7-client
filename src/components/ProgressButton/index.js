// @flow
import React, { PureComponent } from 'react';
import T from 'prop-types';

import classname from 'classnames/bind';

import style from './style.css';

import save from 'images/icons/ico-ebi-save.svg';
import download from 'images/icons/ico-ebi-download.svg';

const s = classname.bind(style);

const CENTER = 60;
const RADIUS = 48;
const STROKE_WIDTH = CENTER - RADIUS;
const CIRCUMFERENCE = 2 * Math.PI * RADIUS;

/*:: type Props = {
  downloading: boolean,
  success: boolean,
  failed: boolean,
  progress: number,
}; */

class ProgressButton extends PureComponent /*:: <Props> */ {
  static propTypes = {
    downloading: T.bool.isRequired,
    success: T.bool.isRequired,
    failed: T.bool.isRequired,
    progress: T.number.isRequired,
  };

  render() {
    const { downloading, success, progress, failed } = this.props;
    return (
      <span className={s('container', { failed })}>
        <svg
          width="3em"
          height="3em"
          viewBox={`0 0 ${2 * CENTER} ${2 * CENTER}`}
          className={s('svg')}
        >
          <g className={s('group', { downloading })}>
            <circle
              cx={CENTER}
              cy={CENTER}
              transform={`rotate(-90 ${CENTER} ${CENTER})`}
              r={RADIUS}
              fill="none"
              className={s('progress')}
              strokeWidth={STROKE_WIDTH}
              strokeLinecap="round"
              strokeDasharray={CIRCUMFERENCE}
              strokeDashoffset={CIRCUMFERENCE * (1 - progress)}
            />
          </g>
          <circle
            cx={CENTER}
            cy={CENTER}
            // circle inside the progress circle (+1 to account for interpolat.)
            r={RADIUS - STROKE_WIDTH / 2 + 1}
            fill="none"
          />
          <g className={s('card', { flipped: success })}>
            <image
              x="30"
              y="10"
              width="60"
              height="100"
              className={s('back')}
              href={download}
            />

            <image
              x="30"
              y="10"
              width="60"
              height="100"
              className={s('front')}
              href={save}
            />
          </g>
        </svg>
      </span>
    );
  }
}

export default ProgressButton;
