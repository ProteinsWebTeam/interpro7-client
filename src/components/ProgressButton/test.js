// @flow
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import ProgressButton from '.';

const renderer = new ShallowRenderer();

describe('<ProgressButton />', () => {
  test('should render, first step', () => {
    renderer.render(
      <ProgressButton
        downloading
        success={false}
        failed={false}
        progress={0}
      />,
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  test('should render, with progress', () => {
    renderer.render(
      <ProgressButton
        downloading
        success={false}
        failed={false}
        progress={0.5}
      />,
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  test('should render, with success', () => {
    renderer.render(
      <ProgressButton
        downloading={false}
        success
        failed={false}
        progress={1}
      />,
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  test('should render, with failure', () => {
    renderer.render(
      <ProgressButton downloading={false} success failed progress={0.8} />,
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
