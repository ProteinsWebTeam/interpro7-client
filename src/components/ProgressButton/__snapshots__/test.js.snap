// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressButton /> should render, first step 1`] = `
<React.Fragment>
  <span
    className="container"
  >
    <svg
      className="svg"
      height="2em"
      viewBox="0 0 180 180"
      width="2em"
    >
      <g
        className="group downloading"
      >
        <circle
          className="progress"
          cx={90}
          cy={90}
          fill="none"
          opacity={1}
          r={70}
          strokeDasharray={439.822971502571}
          strokeDashoffset={439.822971502571}
          strokeLinecap="round"
          strokeWidth={20}
          transform="rotate(-90 90 90)"
        />
        <circle
          cx={90}
          cy={90}
          fill="none"
          r={61}
        />
      </g>
    </svg>
    <span
      className="download-icon icon icon-common icon-download-alt"
    />
  </span>
</React.Fragment>
`;

exports[`<ProgressButton /> should render, with failure 1`] = `
<React.Fragment>
  <span
    className="container failed"
  >
    <svg
      className="svg"
      height="2em"
      viewBox="0 0 180 180"
      width="2em"
    >
      <g
        className="group"
      >
        <circle
          className="progress"
          cx={90}
          cy={90}
          fill="none"
          opacity={0}
          r={70}
          strokeDasharray={439.822971502571}
          strokeDashoffset={87.96459430051418}
          strokeLinecap="round"
          strokeWidth={20}
          transform="rotate(-90 90 90)"
        />
        <circle
          cx={90}
          cy={90}
          fill="none"
          r={61}
        />
      </g>
    </svg>
    <span
      className="download-icon icon icon-common icon-save"
    />
  </span>
</React.Fragment>
`;

exports[`<ProgressButton /> should render, with progress 1`] = `
<React.Fragment>
  <span
    className="container"
  >
    <svg
      className="svg"
      height="2em"
      viewBox="0 0 180 180"
      width="2em"
    >
      <g
        className="group downloading"
      >
        <circle
          className="progress"
          cx={90}
          cy={90}
          fill="none"
          opacity={1}
          r={70}
          strokeDasharray={439.822971502571}
          strokeDashoffset={219.9114857512855}
          strokeLinecap="round"
          strokeWidth={20}
          transform="rotate(-90 90 90)"
        />
        <circle
          cx={90}
          cy={90}
          fill="none"
          r={61}
        />
      </g>
    </svg>
    <span
      className="download-icon icon icon-common icon-download-alt"
    />
  </span>
</React.Fragment>
`;

exports[`<ProgressButton /> should render, with success 1`] = `
<React.Fragment>
  <span
    className="container"
  >
    <svg
      className="svg"
      height="2em"
      viewBox="0 0 180 180"
      width="2em"
    >
      <g
        className="group"
      >
        <circle
          className="progress"
          cx={90}
          cy={90}
          fill="none"
          opacity={0}
          r={70}
          strokeDasharray={439.822971502571}
          strokeDashoffset={0}
          strokeLinecap="round"
          strokeWidth={20}
          transform="rotate(-90 90 90)"
        />
        <circle
          cx={90}
          cy={90}
          fill="none"
          r={61}
        />
      </g>
    </svg>
    <span
      className="download-icon icon icon-common icon-save"
    />
  </span>
</React.Fragment>
`;
